File: services-turf/src/test/java/com/trackasia/turf/TurfTransformationTest.java
package com.trackasia.turf;

import com.trackasia.geojson.Feature;
import com.trackasia.geojson.FeatureCollection;
import com.trackasia.geojson.Polygon;
import com.trackasia.geojson.Point;

import org.junit.Ignore;
import org.junit.Test;

public class TurfTransformationTest extends TestUtils {

  private static final String CIRCLE_IN = "turf-transformation/circle_in.json";
  private static final String CIRCLE_OUT = "turf-transformation/circle_out.json";

  @Test
  @Ignore
  public void name() throws Exception {
    Feature feature = Feature.fromJson(loadJsonFixture(CIRCLE_IN));
    Polygon polygon = TurfTransformation.circle((Point) feature.geometry(),
      feature.getNumberProperty("radius").doubleValue());

    FeatureCollection featureCollection = FeatureCollection.fromJson(loadJsonFixture(CIRCLE_OUT));
    compareJson(featureCollection.features().get(1).geometry().toJson(), polygon.toJson());
  }
}

File: services-turf/src/commonTest/kotlin/com/trackasia/geojson/turf/TurfTransformationTest.kt
package com.trackasia.geojson.turf

import kotlin.test.Test
import com.trackasia.geojson.model.Feature
import com.trackasia.geojson.model.Point
import com.trackasia.geojson.model.Polygon
import com.trackasia.geojson.turf.TestUtils.loadJsonFixture
import com.trackasia.geojson.turf.TurfTransformation.circle
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class TurfTransformationTest {

    @Test
    fun throwOnInvalidSteps() {
        val featureIn = Feature.fromJson(loadJsonFixture(CIRCLE_IN))

        assertFailsWith(IllegalArgumentException::class) {
            circle(
                center = featureIn.geometry as Point,
                steps = -1,
                radius = featureIn.getDoubleProperty("radius")!!
            )
        }

        assertFailsWith(IllegalArgumentException::class) {
            circle(
                center = featureIn.geometry as Point,
                steps = 0,
                radius = featureIn.getDoubleProperty("radius")!!
            )
        }
    }

    @Test
    fun pointToCircle() {
        val featureIn = Feature.fromJson(loadJsonFixture(CIRCLE_IN))
        val circlePolygon = circle(
            center = featureIn.geometry as Point,
            radius = featureIn.getDoubleProperty("radius")!!
        )

        val expectedPolygonFeature = Feature.fromJson(loadJsonFixture(CIRCLE_OUT))
        val expectedPolygon = expectedPolygonFeature.geometry as Polygon

        assertEquals(circlePolygon.coordinates.size, expectedPolygon.coordinates.size)
        for (i in circlePolygon.coordinates.indices) {
            val circleLineCoordinates = circlePolygon.coordinates[i]
            val expectedLineCoordinates = expectedPolygon.coordinates[i]

            assertEquals(circleLineCoordinates.size, expectedLineCoordinates.size)
            for (x in circleLineCoordinates.indices) {
                val circlePoint = circleLineCoordinates[x]
                val expectedPoint = expectedLineCoordinates[x]

                assertEquals(circlePoint.longitude, expectedPoint.longitude, TestUtils.DELTA)
                assertEquals(circlePoint.latitude, expectedPoint.latitude, TestUtils.DELTA)
            }
        }
    }

    companion object {
        private const val CIRCLE_IN = "turf-transformation/circle_in.json"
        private const val CIRCLE_OUT = "turf-transformation/circle_out.json"
    }
}
